File: server/worker-auth/src/services/transformers/user.transformer.ts
import { User, UserSearchDocument, UserViewModel } from '../../models';

export class UserTransformer {
  static getDocumentId(organization: string, userId: string, visibility: string): string {
    return `${organization}_${userId.split('|')[0]}_${visibility}`;
  }

  static toViewModel(user: User, visibility: string, roles: string[]): UserViewModel {
    const showExternal = user.user_metadata?.showExternally ?? [];
    return {
      userId: user.user_id,
      fullName: user.name,
      email: user.email,
      picture: user.picture,
      phone: user.phone_number,
      createdAt: user.created_at,
      updatedAt: user.updated_at,
      lastLogin: user.last_login,
      loginCount: user.logins_count,
      title: visibility === 'private' || showExternal.includes('title') ? user.user_metadata?.title : 'private',
      linkedIn: visibility === 'private' || showExternal.includes('linkedIn') ? user.user_metadata?.linkedin : 'private',
      twitter: visibility === 'private' || showExternal.includes('twitter') ? user.user_metadata?.twitter : 'private',
      roles,
    };
  }
}

File: server/core/Services/Transformers/UserTransformer.cs
using Auth0.ManagementApi.Models;
using Wbs.Core.DataServices;
using Wbs.Core.ViewModels;

namespace Wbs.Core.Services.Transformers;

public class UserTransformer : SqlHelpers
{
    public static List<UserViewModel> ToViewModelList(IEnumerable<User> users)
    {
        var results = new List<UserViewModel>();

        foreach (var user in users)
            results.Add(ToViewModel(user));

        return results;
    }

    public static UserViewModel ToViewModel(User user)
    {
        return new UserViewModel
        {
            CreatedAt = user.CreatedAt,
            UpdatedAt = user.UpdatedAt,
            Email = user.Email,
            FullName = user.FullName,
            LastLogin = user.LastLogin,
            LinkedIn = user.UserMetadata?.linkedIn,
            LoginCount = int.Parse(user.LoginsCount ?? "0"),
            Phone = user.PhoneNumber,
            Picture = user.Picture,
            Title = user.UserMetadata?.title,
            Twitter = user.UserMetadata?.twitter,
            UserId = user.UserId,
            ShowExternally = user.UserMetadata?.showExternally == null ? [] : user.UserMetadata?.showExternally.ToObject<string[]>(),
        };
    }

    public static UserUpdateRequest ToUpdateUserRequest(UserViewModel viewModel)
    {
        return new UserUpdateRequest
        {
            AppMetadata = new { },
            UserMetadata = new
            {
                title = viewModel.Title,
                linkedIn = viewModel.LinkedIn,
                twitter = viewModel.Twitter,
                showExternally = viewModel.ShowExternally,
            },
            FullName = viewModel.FullName
        };
    }
}